### PR TITLE
opponentplay

### PR DESCRIPTION
**Pull request by lemz1**
Good afternoon,
I need to repost this pull request (#2770), since I was kinda stupid, but it isn't just a repost.
I updated it to the new code and changed some stuff to have less variables and less lines.

### What does this do ###
This adds the gameplay option 'opponentplay'. If turned on lets you play the opponent chart.
It isn't a big feature but yeah.

### CODE REVIEW

**Comment by tposejank** [comment]
sick

**Comment by HiroMizuki** [comment]
sick

**Comment by CrusherMods** [comment]
sick

**Comment by MayoALT** [comment]
sick

**Comment by SrtHero278** [comment]
sick

**Comment by MasterDirector99** [comment]
sick

**Comment by JorgeMiguel22** [comment]
why it doesnt appear anything about opponent play?

**Comment by HiroMizuki** [comment]
> why it doesnt appear anything about opponent play?

what

**Comment by JorgeMiguel22** [comment]
i searched in the options and didnt find the opponent play option

**Comment by HiroMizuki** [comment]
> i searched in the options and didnt find the opponent play option

are u sure that u coded everything changed at the PR?
or are u sure that u compiled the right source?

**Comment by JorgeMiguel22** [comment]
yeah i changed all the files needed

**Comment by SrtHero278** [comment]
If you're looking through the settings through the main menu. That's not how you find it. You gotta select a song and press `control`.

**Comment by JorgeMiguel22** [comment]
oh it worked, thx for helping

**Comment by Megalonumber0ne** [comment]
I coded everything in correctly but when i went to play a song (I'll use Dadbattle for an example), all of the notes appear on the opponent's side, including BF's, also when BF and DD sync both of the notes appear at the same time and can only be pressed by either the AI or you (It's mostly the AI that gets the notes)
image
image

**Comment by lemz1** [comment]
> I coded everything in correctly but when i went to play a song (I'll use Dadbattle for an example), all of the notes appear on the opponent's side, including BF's, also when BF and DD sync both of the notes appear at the same time and can only be pressed by either the AI or you (It's mostly the AI that gets the notes) image image
> 
> EDIT: I have a screenshot of the notes mixing image

The note generation is somewhere in line 1870.
Are you sure that you put in
```
if (songNotes[1] <= 3 && opponentChart)
{
	gottaHitNote = !section.mustHitSection;
}
```
Maybe you did a wrong symbol instead of `<=`

**Comment by Megalonumber0ne** [comment]
Even after checking all of the code that was given to me and re-building the game i still got the same result

**Comment by Megalonumber0ne** [comment]
Uploading Vide1.mp4…

**Comment by Gui-iago** [comment]
nice! merged this into my fork and it works perfectly, I've seen some bugs of course (some of which I managed to fix) but here are all of the bugs I could find

* since Dad is now BF when Opponent Mode is on, Lua Scripts are completely swapped and can present issues (such as animations not working properly if lua scripts are present, or noteskins being swapped, etc)
* Opponent Mode Dad can't hold notes, his animation stops looping like it should if he was actually Player 2 if you try to do so, and it loops BF's animation instead
* GF Notes make it so Dad **and** GF sing together
  Solution: on line 4013, add "&& !gfNote" in "if(opponentChart) {"
  
* Alt Anims and Hey Notes are completely bugged, or straight up don't work sometimes (try playing Cocoa or Eggnog)
* sometimes (rarely) the charts don't swap, nothing major
* Miss Animations won't play if Ghost Tapping is disabled for Player 2 (try any Week 3 song with Ghost Tapping off)

that everything I know!

**Comment by XtraXD1** [comment]
sick

**Comment by Gui-iago** [comment]
(latest update)


https://user-images.githubusercontent.com/45212377/150021456-a5daa0f5-11f9-4929-a9bd-0b3145fc5f4c.mp4


https://user-images.githubusercontent.com/45212377/150021880-62782d0d-c79b-424d-807b-e5dc6835fbec.mp4

the same doesn't happen for BF

**Comment by lemz1** [comment]
> (latest update)
> 2022-01-18.17-27-08.mp4
> 2022-01-18.17-30-26.mp4
> 
> the same doesn't happen for BF

Are you sure you coded everything right. The thing with bf dying doesn't really make sense, since I didn't change how health works.

**Comment by Gui-iago** [comment]
> > (latest update)
> > 2022-01-18.17-27-08.mp4
> > 2022-01-18.17-30-26.mp4
> > the same doesn't happen for BF
> 
> Are you sure you coded everything right. The thing with bf dying doesn't really make sense, since I didn't change how health works. Also I noticed that you are using "Bedrock Engine". There's probably different code

the latest commit fixed it

**Comment by Stilic** [comment]
> Also I noticed that you are using "Bedrock Engine".

It is his own fork of Pysch Engine with custom features from different PRs.

**Comment by Gui-iago** [comment]
BE doesn't change the code as much, it's just that I probably didn't merge the changes correctly, yet it worked with the latest commit fine

**Review by crowplexus — APPROVED** [review]
epic

**Review by stilic — APPROVED** [review]
so epic than i will shit from fear if @ShadowMario or @bbpanzu doesn't merge this pr.

**Review by HiroMizuki — APPROVED** [review]

**Review by crowplexus — APPROVED** [review]

**Comment by Gui-iago** [comment]
image
this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error

**Comment by lemz1** [comment]
> image this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error

Does it tell you which file is the problem? If yes then I can probably help else it will be harder

**Comment by Gui-iago** [comment]
> > image this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error
> 
> Does it tell you which file is the problem? If yes then I can probably help else it will be harder

no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off

**Comment by Gui-iago** [comment]
just checked, also happens if botplay is on

**Comment by lemz1** [comment]
> 

> > > image this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error
> > 
> > 
> > Does it tell you which file is the problem? If yes then I can probably help else it will be harder
> 
> no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off

Can you give me your code like upload it to drive or if you have it in github can you tell me which repository.
I can then thoroughly look at it

**Comment by Gui-iago** [comment]
> > no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off
> 
> Can you give me your code like upload it to drive or if you have it in github can you tell me which repository. I can then thoroughly look at it
https://github.com/Gui-iago/FNF-BedrockEngine/commit/b178ebbfa5a3dbe658da2589440fa5538d2c36ce


https://user-images.githubusercontent.com/45212377/152657238-21654b8e-b80b-4f58-aa83-355b97928b82.mp4

the fixes probably work though, as bf's loops are shorter

**Comment by lemz1** [comment]
I think you need to remove in line 4388 
**_{_**

**Comment by Gui-iago** [comment]
figured it out, it was an issue with Conductor and json shit, rather than your fork

keep up the good work!

**Comment by lemz1** [comment]
> figured it out, it was an issue with Conductor and json shit, rather than your fork

nice, I am glad you figured it out, because I was being too stupid to figure it out

> keep up the good work!

I'll try

**Comment by lemz1** [comment]
[This](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a) needs to be turned of depending of what goodNoteHit does.
For example it needs to be turned off if you use something like,
```
function goodNoteHit()

	accuracyUpdateShitOrSomethingLikeThat()

end
```
since the accuracy update needs to be done everytime the note is hit.
It doesn't need to be turned off if you use something like,
```
function goodNoteHit()

	debugPrint('bf sings')

end
```
since it is supposed to activate when bf hits a note.

**Comment by lemz1** [comment]
Please tell me if there are any bugs or crashes because of [this](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a)

**Comment by tposejank** [comment]
> [This](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a) needs to be turned off depending on what goodNoteHit or opponentNoteHit does. For example it needs to be turned off if you use something like,
> 
> ```
> function goodNoteHit()
> 
> 	accuracyUpdateShitOrSomethingLikeThat()
> 
> end
> ```
> 
> since the accuracy update needs to be done everytime the note is hit. It can be turned off by doing this
> 
> ```
> function onCreate()
> 
>     setProperty('noteHitFix', false)
> 
> end
> ```
> 
> It doesn't need to be turned off if you use something like,
> 
> ```
> function goodNoteHit()
> 
> 	debugPrint('bf sings')
> 
> end
> ```
> 
> since it is supposed to activate when bf hits a note.

https://c.tenor.com/DuThn51FjPcAAAAC/nerd-emoji-nerd.gif

**Comment by frantastic24** [comment]
having botplay and opponent chart turned on at the same time seems to make the opponent not reset back to idle after singing

**Review by JorosTheModderPH — APPROVED** [review]

**Review by JorosTheModderPH — APPROVED** [review]

**Comment by Gui-iago** [comment]
there's a glitch with the Cam Zoom event, but I found the solution pretty quickly

image

**Review by FletchunterMC108338 — APPROVED** [review]

**Comment by FletcherMC0329** [comment]
Nice but the only problem is lua swapped when opponentplay.

**Comment by FletcherMC0329** [comment]
Ok

File: source/GameplayChangersSubstate.hx [hunk 1]
```diff
@@ -91,7 +91,7 @@ class GameplayChangersSubstate extends MusicBeatSubstate
 		var option:GameplayOption = new GameplayOption('Botplay', 'botplay', 'bool', false);
 		optionsArray.push(option);
 
-		var option:GameplayOption = new GameplayOption('Play Opponents Chart', 'opponentplay', 'bool', false);
+		var option:GameplayOption = new GameplayOption("Play Opponent's Chart", 'opponentplay', 'bool', false);
```
Comment (Gui-iago) [inline source/GameplayChangersSubstate.hx:94]: I think a better name for this could be "Play as Opponent"

Comment (lemz1) [reply source/GameplayChangersSubstate.hx:94]: shish kebab

**Comment by frantastic24** [comment]
this seems to break gf characters singing when the option is on, they don't return to their idle

https://user-images.githubusercontent.com/95725102/156490750-f4516eef-ebef-4801-822f-e83014d32899.mp4

**Comment by BeastlyGhost** [comment]
oh yeah since this is closed and finished now and won't be merged, I recommend to everyone who wants this feature to simply merge them to your own forks or private builds

**Pull request closed without merging**